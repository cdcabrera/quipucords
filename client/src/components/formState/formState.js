import React from 'react';
import PropTypes from 'prop-types';
import _cloneDeep from 'lodash/cloneDeep';
import _isEqual from 'lodash/isEqual';
import helpers from '../../common/helpers';

class FormState extends React.Component {
  constructor(props) {
    super(props);

    let refValues = null;

    if (props.resetRefValues === true) {
      refValues = _cloneDeep(props.initialValues);
    }

    this.state = {
      errors: {},
      isUpdating: false,
      isSubmitting: false,
      isValidating: false,
      isValid: null,
      refValues,
      submitCount: 0,
      touched: {},
      values: _cloneDeep(props.initialValues)
    };
  }

  componentDidMount() {
    const { validateOnmount } = this.props;

    if (validateOnmount === true) {
      this.validateOnMount();
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const { initialValues, resetRefValues } = this.props;

    if (resetRefValues === true && !_isEqual(prevState.refValues, initialValues)) {
      this.updateComponent();
    }
  }

  onEventCustom = custom => {
    if (!custom.name || !custom.value) {
      return;
    }

    this.onEvent({ target: { ...custom }, persist: helpers.noop, type: 'custom' });
  };

  onEvent = event => {
    const { touched, values } = this.state;
    const { id, name, value } = event.options ? { ...event } : event.target;

    event.persist();

    const targetName = name || id || 'generated form state target, add name or id attr to field';
    const updatedTouched = { ...touched, [targetName]: true };
    const updatedValues = { ...values, [targetName]: value };

    this.setState(
      {
        values: updatedValues,
        touched: updatedTouched,
        isUpdating: true,
        isValidating: true
      },
      () =>
        this.validate(event).then(updatedErrors => {
          const setErrors = { ...((updatedErrors && updatedErrors[0]) || updatedErrors || {}) };
          const checkIsValid = !Object.keys(setErrors).length;

          this.setState({
            errors: setErrors,
            isUpdating: false,
            isValidating: false,
            isValid: checkIsValid
          });
        })
    );
  };

  onReset = event => {
    const { refValues, values } = this.state;
    const { onReset, resetRefValues } = this.props;

    event.persist();

    if (refValues && resetRefValues === true) {
      const updatedValues = _cloneDeep(refValues);

      this.setState(
        {
          touched: {},
          submitCount: 0,
          values: updatedValues
        },
        () => onReset({ event, values: updatedValues, prevValues: values })
      );
    } else {
      onReset({ event, values });
    }
  };

  onSubmit = event => {
    const { submitCount } = this.state;

    event.persist();
    event.preventDefault();

    this.setState(
      {
        submitCount: submitCount + 1,
        isSubmitting: true,
        isUpdating: true,
        isValidating: true
      },
      () =>
        this.validate(event).then(updatedErrors => {
          const setErrors = { ...((updatedErrors && updatedErrors[0]) || updatedErrors || {}) };
          const checkIsValid = !Object.keys(setErrors).length;

          this.setState(
            {
              errors: setErrors,
              isSubmitting: false,
              isUpdating: false,
              isValid: checkIsValid,
              isValidating: false,
              touched: {}
            },
            () => !Object.keys(updatedErrors).length && this.submit(event)
          );
        })
    );
  };

  submit(event) {
    const { values } = this.state;
    const { onSubmit } = this.props;

    return Promise.resolve(onSubmit({ event, values }));
  }

  validate(event) {
    const { values } = this.state;
    const { validate } = this.props;

    const checkPromise = validate({ event, values });

    if (Object.prototype.toString.call(checkPromise) === '[object Promise]') {
      return checkPromise;
    }

    return {
      then: callback => callback(checkPromise)
    };
  }

  validateOnMount(event) {
    const updatedEvent = event || {
      type: 'mount'
    };

    this.setState(
      {
        isUpdating: true,
        isValidating: true
      },
      () =>
        this.validate(updatedEvent).then(updatedErrors => {
          const setErrors = { ...((updatedErrors && updatedErrors[0]) || updatedErrors || {}) };
          const checkIsValid = !Object.keys(setErrors).length;

          this.setState({
            errors: setErrors,
            isUpdating: false,
            isValidating: false,
            isValid: checkIsValid
          });
        })
    );
  }

  updateComponent() {
    const { initialValues } = this.props;

    this.setState(
      {
        isUpdating: true
      },
      () => {
        const values = _cloneDeep(initialValues);
        this.setState({
          isUpdating: false,
          refValues: initialValues,
          values
        });
      }
    );
  }

  render() {
    const { children } = this.props;

    return (
      <React.Fragment>
        {children({
          handleOnEventCustom: this.onEventCustom,
          handleOnEvent: this.onEvent,
          handleOnReset: this.onReset,
          handleOnSubmit: this.onSubmit,
          ...this.state
        })}
      </React.Fragment>
    );
  }
}

FormState.propTypes = {
  children: PropTypes.func.isRequired,
  initialValues: PropTypes.object,
  onReset: PropTypes.func,
  onSubmit: PropTypes.func,
  resetRefValues: PropTypes.bool,
  validate: PropTypes.func,
  validateOnmount: PropTypes.bool
};

FormState.defaultProps = {
  initialValues: {},
  onReset: helpers.noop,
  onSubmit: helpers.noop,
  resetRefValues: false,
  validate: helpers.noop,
  validateOnmount: false
};

export { FormState as default, FormState };
