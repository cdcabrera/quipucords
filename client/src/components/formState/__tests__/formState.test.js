import React from 'react';
import { mount } from 'enzyme';
import { FormState } from '../formState';

describe('FormState Component', () => {
  it('should render a basic component', () => {
    const props = { resetRefValues: true, initialValues: { lorem: 'ipsum' } };

    const component = mount(
      <FormState {...props}>
        {({ values }) => (
          <form>
            <label>
              Lorem
              <input name="lorem" value={values.lorem} readOnly type="text" />
            </label>
          </form>
        )}
      </FormState>
    );

    const componentInstance = component.instance();
    expect(componentInstance.props).toMatchSnapshot('initial props');
    expect(componentInstance.state).toMatchSnapshot('initial state');

    expect(component.render()).toMatchSnapshot('basic render');
  });

  it('should update handle change and submits events while updating state', () => {
    const props = { initialValues: { lorem: 'ipsum' }, validate: () => ({}) };

    const component = mount(
      <FormState {...props}>
        {({ values, handleOnEvent, handleOnSubmit }) => (
          <form onSubmit={handleOnSubmit}>
            <label>
              Lorem
              <input name="lorem" value={values.lorem} type="text" onChange={handleOnEvent} />
            </label>
            <button type="submit">Submit</button>
          </form>
        )}
      </FormState>
    );

    const componentInstance = component.instance();
    const mockEvent = { target: { value: 'dolor', name: 'lorem' }, persist: () => {}, preventDefault: () => {} };
    componentInstance.onEvent(mockEvent);
    expect(component.state()).toMatchSnapshot('onevent');

    componentInstance.onSubmit({ persist: () => {}, preventDefault: () => {} });
    expect(component.state()).toMatchSnapshot('onsubmit');
  });

  it('should do a basic validation check', () => {
    const props = {
      initialValues: {
        lorem: 'ipsum'
      },
      validate: ({ values }) => {
        const updatedErrors = {};

        if (!values.lorem) {
          updatedErrors.lorem = 'required';
        }

        return updatedErrors;
      }
    };

    const component = mount(
      <FormState {...props}>
        {({ errors, values, handleOnEvent }) => (
          <form>
            <label>
              Lorem
              <input id="lorem" value={values.lorem} type="text" onChange={handleOnEvent} />
              <span className="error">{errors.lorem}</span>
            </label>
          </form>
        )}
      </FormState>
    );

    const mockEvent = { target: { value: '', id: 'lorem' }, persist: () => {}, preventDefault: () => {} };
    component.instance().onEvent(mockEvent);
    expect(component.state()).toMatchSnapshot('basic validation');
  });

  it('should handle custom events', () => {
    const props = {
      initialValues: {
        lorem: 'ipsum',
        dolor: ''
      },
      validate: ({ values }) => {
        const updatedErrors = {};

        if (!values.lorem) {
          updatedErrors.lorem = 'required';
        }

        return updatedErrors;
      }
    };

    const component = mount(
      <FormState {...props}>
        {({ errors, values, handleOnEventCustom }) => (
          <form>
            <label>
              Lorem
              <input name="lorem" value={values.lorem} type="text" onChange={handleOnEventCustom} />
              <span className="error">{errors.lorem}</span>
              <input name="dolor" value={values.dolor} type="hidden" />
            </label>
          </form>
        )}
      </FormState>
    );

    component.instance().onEventCustom({ name: 'lorem', value: 'woot' });
    expect(component.state()).toMatchSnapshot('single custom event');

    component.instance().onEventCustom([{ name: 'lorem', value: 'woot again' }, { name: 'dolor', value: 'sit' }]);
    expect(component.state()).toMatchSnapshot('multiple custom events');
  });
});
