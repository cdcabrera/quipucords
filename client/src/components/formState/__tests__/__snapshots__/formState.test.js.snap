// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic render 1`] = `
<form>
  <label>
    Lorem
    <input
      name="lorem"
      readonly=""
      type="text"
      value="ipsum"
    />
  </label>
</form>
`;

exports[`basic validation 1`] = `
Object {
  "lorem": "required",
}
`;

exports[`initial props 1`] = `
Object {
  "allowUpdates": false,
  "children": [Function],
  "initialValues": Object {
    "lorem": "ipsum",
  },
  "initialValuesAssumeBoolIsChecked": true,
  "onReset": [Function],
  "onSubmit": [Function],
  "resetRefValues": true,
  "resetUsingInitialValues": false,
  "validate": [Function],
  "validateOnmount": false,
}
`;

exports[`initial state 1`] = `
Object {
  "isSubmitting": false,
  "isUpdating": false,
  "isValid": null,
  "isValidating": false,
  "submitCount": 0,
}
`;

exports[`multiple custom events 1`] = `
Object {
  "dolor": "sit",
  "lorem": "woot again",
}
`;

exports[`not mutated 1`] = `
Object {
  "lorem": "ipsum",
}
`;

exports[`onevent 1`] = `
Object {
  "isSubmitting": false,
  "isUpdating": false,
  "isValid": true,
  "isValidating": false,
  "submitCount": 0,
}
`;

exports[`onevent values updated 1`] = `
Object {
  "lorem": "dolor",
}
`;

exports[`onreset 1`] = `
Object {
  "isSubmitting": false,
  "isUpdating": false,
  "isValid": null,
  "isValidating": false,
  "submitCount": 0,
}
`;

exports[`onsubmit 1`] = `
Object {
  "isSubmitting": false,
  "isUpdating": false,
  "isValid": true,
  "isValidating": false,
  "submitCount": 1,
}
`;

exports[`reset values updated 1`] = `
Object {
  "lorem": "ipsum",
}
`;

exports[`single custom event 1`] = `
Object {
  "dolor": "",
  "lorem": "woot",
}
`;
